### ISSUE TITLE
[Astro] `createResource` results get mixed up when rendering server only components with Suspense

### ISSUE THREAD

**Issue by PeterDraex** [solidjs/solid#2131]
### Describe the bug

`resource.data()` is returning data from another resource, in another component

### Your Example Website or App

https://stackblitz.com/edit/github-pqy8j7-y8z78u?file=src%2Fcomponents%2FPage.tsx

### Steps to Reproduce the Bug or Issue

1. Open Page.tsx
2. See that `pageQuery` resolves to `'pageQuery API response'` in it's fetcher function, but when you check `pageQuery.data()`, it contains `"Component A API response"`

### Expected behavior

This text is displayed
> pageQuery.data: "pageQuery API response"

### Platform

Windows, Chrome 123

### Additional context

@birkskyum:
> It appear to happen because the data from the CompA resource isn't used anywhere, and then it ends up being picked up inside Page instead of the resource defined later there...

**Comment by ryansolid** [comment]
The setup for this is really odd. In that Astro seems to SSR Solid twice at least in dev and that is causing a whole bunch of chaos (not to mention sort of terrible for performance). I'm going to try to comment out the Astro code and see if there is any issue on Solid's side.

**Comment by PeterDraex** [comment]
> Astro seems to SSR Solid twice

Unfortunately, that's by design according to https://github.com/withastro/astro/issues/8368
> Unfortunately this is because we can't detect if a given function is a Solid component until we run it. Our React and Preact integrations do the same thing.

**Comment by ryansolid** [comment]
Ok positive is I don't think this is the cause even if it causes some weirdness in terms of multiple copies of resources. I will continue to investigate.

**Comment by ryansolid** [comment]
Ok I see the problem I just need to decide where the fix goes. There is an optimization it looks like someone put in to Solid-Astro integration to generate the non-hydration markup when you don't add client directives to your Solid components. NoHydration doesn't increment component IDs and parallel Suspense is resetting at the same id making those resources have identical IDs.

I admit I wasn't expecting this optimization. But I think this could happen in Solid's own Island implementation so I guess I need to make sure Suspense generates its own unique IDs. It will add depth but I'm not sure how else to avoid it.

**Comment by PeterDraex** [comment]
Hi, can I ask if this will be fixed in foreseeable future? Or is there any other way to avoid sending serialized data to the browser?

**Comment by ryansolid** [comment]
It's tricky to fix. I'd kinda rather see the Astro adapter drop the optimization in the short term rather than try to introduce new alternative functionality into Solid. Not serializing the data means it can't hydrate. I expected most usage of Astro to fetch on the server and use `createResource` source as initial with `initialValue` but if this is a Tanstack query thing then that probably doesn't make sense. I'm doing a release today and timeboxing it. I've been looking at this for the last hour if I can solve it in the next it goes in it, otherwise it will take some time.

**Comment by ryansolid** [comment]
Oh I guess I should mention there is a no scripts option in our render function that I think could be turned on in the no hydration case like yours but that isn't what is causing this issue and it would need to be exposed or set by the Astro renderer. 

The problem here is id collision during SSR. I can fix it by removing an optimization to reduce id depth with no-hydration.. basically reducing the efficiency of the no-hydration optimization but I think it will work. I will need to test it though in other SSR environments.

**Comment by ryansolid** [comment]
It will work but it would break SolidStart so I can't make this a patch release. It changes the SSR behavior of libraries depending on `<NoHydrate>`. They will have to account for the change in ID generation. So I can't fix this now.

I think that if you are in the no hydration scenario fetch your data outside of Solid and pass it through is my recommendation. There is nothing interactive so the data fetching doesn't need to live in Solid. All the other cases seem to work.

**Comment by elite174** [comment]
https://github.com/withastro/astro/pull/12131

This could be a fix for one renderer.

**Comment by TomzBench** [comment]
my resources seem to be getting mixed up and i have no idea why. Is there any gotchas here that can happen (unrelated to astro?)

**Comment by ryansolid** [comment]
> my resources seem to be getting mixed up and i have no idea why. Is there any gotchas here that can happen (unrelated to astro?)

Yeah much less likely than the `<NoHydration>` issue. But given the way the ids generate is possible for branching logic to cause collisions. `createResource` on the server because of the way we re-render currently has to basically follow the something similar to React's hook rules. Because to match it up we use the reactive/component depth and the order it is declared in. Since all decision points in the JSX are known we can ensure stuff loading in after the fact keep stable IDs but.. it is possible if there is `isServer` or conditional logic wrapping `createResource` calls or if conditional branching exactly mirrors the alternative that a mismatch could happen. That being said I haven't seen any of these cases in the wild since generally people don't do that.

More than likely though it is caused by things rendering multiple times, like accessing `props.children` multiple times causing some oddness.

**Comment by TomzBench** [comment]
Thanks for detailed reply. It turns out my fetch calls were returning out of order. and the mixing up was happening by the browser.

So if the responses returned in a different order than the requests were made, the mixup would happen there.  I did not expect that.

**Comment by ryansolid** [comment]
> So if the responses returned in a different order than the requests were made, the mixup would happen there. I did not expect that.

Generally that shouldn't, but if the tree hierarchy looses Ids it is possible to lose the heirarchy. We record all decision points at a fixed id. Which means if stuff happens out of order the decision is atleast fixed. Children have their own hierarchical ids and don't collide for that reason. So unless you are conditionally choosing where to createResource, then this is very unlikely. More than likely there is some other bug in the code causing the issue.


On the positive since we use the reactive graph to generate IDs unconditionally in Solid 2.0. This won't be an issue. But I have no easy path for 1.0 without messing with the existing ecosystem.

**Comment by ryansolid** [comment]
Won't fix in 1.x. NoHydration skips component id increments, so parallel Suspense can collide resource ids. Changing that id gen would break SolidStart (dummy components matching client/server ids). Already fixed in 2.0. Same resolution as #2546 / #2707.
